[SERVER-2220] Clarification around adding custom Telegraf config (#7675)
circleci/circleci-docs: jekyll/_cci2/server/v4.1/operator/using-metrics.adoc

[#telegraf]
=== Telegraf
Most services running on server report StatsD metrics to the https://www.influxdata.com/time-series-platform/telegraf/[Telegraf] pod running in server.

> The configuration is fully customizable, so you can forward your metrics from Telegraf to any output supported by Telegraf through https://docs.influxdata.com/telegraf/v1.24/plugins/#output-plugins[output plugins]. By default, it will provide a metrics endpoint for Prometheus to scrape.

Author: @circleci/security-controls / Server reviewers:

The (somewhat) big change to the way Telegraf is now being configured (which we don't state anywhere explicitly) is that we are _technically_ no longer allowing the end user to completely replace the default telegraf configuration. This has to do with the changes to [telegraf-config.yaml](https://github.com/circleci/server/blob/main/templates/telegraf-config.yaml) in that the custom configuration is now getting _appended_ to what we ship by [default](https://github.com/circleci/server/blob/a77fd029851c0ccb5752b8d2ca5aae0e9f3c0419/values.yaml#L513-L552).The reason for this has to do with the metrics/usage statistics work where we needed to have sections of Telegraf config that the user shouldn't be able to override (otherwise we would have lost our metrics shipping config).

The **TL;DR** to the above is that an en user can adjust any aspect of Telegraf config except for the default `statsd` input and `prometheus` output plugin config sections. Anything else can be overwritten by suppling Telegraf compatible config yaml in `telgraf.config.custom_config_file`.

Author: I haven't added a note to the documentation about the above ☝️ because I don't think there's a point and it might introduce unnecessary confusion...

> The configuration is fully customizable, so you can forward your metrics from Telegraf to any output supported by Telegraf through https://docs.influxdata.com/telegraf/v1.24/plugins/#output-plugins[output plugins]. By default, it will provide a metrics endpoint for Prometheus to scrape.

Reviewer: ```suggestion
The configuration is fully customizable. Metrics can be forwarded from Telegraf to any output supported by Telegraf through https://docs.influxdata.com/telegraf/v1.24/plugins/#output-plugins[output plugins]. By default, it will provide a metrics endpoint for Prometheus to scrape.
```

[#use-telegraf-to-forward-metrics-to-datadog]
=== Use Telegraf to forward metrics to Datadog

> Telegraf ships configured to collect statsd metrics using the https://docs.influxdata.com/telegraf/v1.24/plugins/#input-statsd[statsd] input plugin and serves these metrics in the Prometheus format on the `/metrics` endpoint using the https://docs.influxdata.com/telegraf/v1.24/plugins/#output-prometheus_client[Prometheus] output plugin. Refer to the `telegraf.config` block in the default `values.yaml` for details on how Telegraf is configured in Server. For details around the different Telegraf configuration options refer to the official https://github.com/influxdata/helm-charts/tree/master/charts/telegraf[Telegraf] Helm chart from Influxdata.

Reviewer 2: ```suggestion
By default, Telegraf is configured to collect statsd metrics using the link:https://docs.influxdata.com/telegraf/v1.24/plugins/#input-statsd[statsd] input plugin, and serves these metrics in the Prometheus format on the `/metrics` endpoint using the https://docs.influxdata.com/telegraf/v1.24/plugins/#output-prometheus_client[Prometheus] output plugin. Refer to the `telegraf.config` block in the default `values.yaml` for details on how Telegraf is configured in CircleCI server. For details on the different Telegraf configuration options, refer to the official link:https://github.com/influxdata/helm-charts/tree/master/charts/telegraf[Telegraf] Helm chart from Influxdata.
```

> As mentioned earlier, Telegraf ships configured to collect statsd metrics using the https://docs.influxdata.com/telegraf/v1.24/plugins/#input-statsd[statsd] input plugin and serves these metrics in Prometheus format on the `/metrics` endpoint using the https://docs.influxdata.com/telegraf/v1.24/plugins/#output-prometheus_client[Prometheus] output plugin. Refer to the `telegraf.config` block in the default `values.yaml` for details on how Telegraf is configured in Server. For details around the different Telegraf configuration options refer to the official https://github.com/influxdata/helm-charts/tree/master/charts/telegraf[Telegraf] Helm chart from Influxdata.

Reviewer: ```suggestion
Telegraf ships configured to collect statsd metrics using the https://docs.influxdata.com/telegraf/v1.24/plugins/#input-statsd[statsd] input plugin and serves these metrics in the Prometheus format on the `/metrics` endpoint using the https://docs.influxdata.com/telegraf/v1.24/plugins/#output-prometheus_client[Prometheus] output plugin. Refer to the `telegraf.config` block in the default `values.yaml` for details on how Telegraf is configured in Server. For details around the different Telegraf configuration options refer to the official https://github.com/influxdata/helm-charts/tree/master/charts/telegraf[Telegraf] Helm chart from Influxdata.
```

Author: Thanks @circlery!

----
...
telegraf:
  args:
    - --config
    - /etc/telegraf/telegraf.d/telegraf_custom.conf
  config:
    custom_config_file: |
      [agent]
        collection_jitter = "0s"
        debug = false
        flush_interval = "10s"
        flush_jitter = "0s"
        hostname = "$HOSTNAME"
        interval = "30s"
        logfile = ""
        metric_batch_size = 1000
        metric_buffer_limit = 10000
        omit_hostname = true
        precision = ""
        quiet = false
        round_interval = true

      [[processors.enum]]
        [[processors.enum.mapping]]
          dest = "status_code"
          field = "status"
          [processors.enum.mapping.value_mappings]
              critical = 3
              healthy = 1
              problem = 2

      [[inputs.statsd]]
        datadog_extensions = true
        metric_separator = "."
        percentile_limit = 1000
        percentiles = [
          50,
          95,
          99
        ]
        service_address = ":8125"
      [[inputs.internal]]
        collect_memstats = false

      [[outputs.file]]
        files = ["stdout"]

      [[outputs.prometheus_client]]
        listen = ":9273"
        path = "/metrics"

> ...

Author: I've decided to remove this sample config yaml block so that we can avoid confusion. We ship Telegraf pre-configured with chart overrides defined [here](https://github.com/circleci/server/blob/a77fd029851c0ccb5752b8d2ca5aae0e9f3c0419/values.yaml#L527). I don't think there's a need to re-iterate these and list them in the docs as the _rendered_ Telegraf yaml config.
